app/rest/metrics-collector/src/main/java/io/syndesis/rest/metrics/collector/MetricsCollector.java

```diff
@@ -0,0 +1,318 @@
+
+import io.fabric8.kubernetes.api.model.Pod;
+import io.fabric8.kubernetes.client.KubernetesClient;
+import io.fabric8.kubernetes.client.internal.readiness.Readiness;
+import io.syndesis.core.Json;
+import io.syndesis.dao.manager.DataManager;
+import io.syndesis.jsondb.GetOptions;
+import io.syndesis.jsondb.JsonDB;
+import io.syndesis.model.ListResult;
+import io.syndesis.model.integration.Integration;
+import io.syndesis.model.metrics.IntegrationMetricsSummary;
+
+@Service
+public class MetricsCollector implements Runnable, Closeable {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(MetricsCollector.class);
+    private static final String HISTORY = "HISTORY";
+
+    private final KubernetesClient kubernetes;
+    private final RawMetricsHandler rawMetricsHandler;
+    private final DataManager dataManager;
+    private final JsonDB jsonDB;
+
+    private final ScheduledExecutorService scheduler = Executors.newSingleThreadScheduledExecutor();
+    private final ExecutorService executor = Executors.newCachedThreadPool(new CustomizableThreadFactory("metrics-collector"));
+
+    @Autowired
+    public MetricsCollector(DataManager dataManager, JsonDB jsonDB, KubernetesClient kubernetes) {
+        this.dataManager = dataManager;
+        this.jsonDB = jsonDB;
+        this.kubernetes = kubernetes;
+        this.rawMetricsHandler = new PersistRawMetrics(jsonDB);
+    }
+
+
+    @PostConstruct
+    @SuppressWarnings("FutureReturnValueIgnored")
+    public void open() {
+        LOGGER.info("Starting metrics collector.");
+        scheduler.scheduleAtFixedRate(this, 10, 5, TimeUnit.SECONDS);
+    }
+
+
+    @Override
+    public void close() throws IOException {
+        LOGGER.info("Stopping metrics collector.");
+        close(scheduler);
+        close(executor);
+    }
+
+
+    @Override
+    public void run() {
+        LOGGER.info("Collecting metrics for active integration pods.");
+        try {
+            List<Pod> integrationPodList = kubernetes.pods().withLabel("integration").list().getItems();
+            Set<String> livePods = new HashSet<>();
+            for (Pod pod : integrationPodList) {
+                livePods.add(pod.getMetadata().getName());
+            }
+
+            integrationPodList
+                .stream()
+                .filter(p -> Readiness.isReady(p))
+                .forEach(p ->
+                    executor.execute(new PodMetricsReader(
+                            kubernetes,
+                            p.getMetadata().getName(),
+                            p.getMetadata().getLabels().get("integration"),
+                            p.getMetadata().getLabels().get("syndesis.io/integration-id"),
+                            p.getMetadata().getLabels().get("syndesis.io/deployment-id"),
+                            rawMetricsHandler))
+            );
+
+            ListResult<Integration> integrationList = dataManager.fetchAll(Integration.class);
```

Comment:
Isn't there an easier way to get all integration ids and names without deserializing all integrations each time the metrics are queried ? Looks super expensive to me. // @chirino

Reply:
@agree: IMHO, we shouldn't even bother about queering the integrations.

I can see that going two ways, none of which requires `fetching all the pods`:

- Record per integration (no rollups)
- Record per reset (with rollups)

# Record per integration

 We should just get all the active pods, grab the data, and save them in the database.

If the object we persisted had a structure like:

    {
      private String integrationId;
      private long lastReset;

      private long messages;  //Messages after lastReset;
      private long errors;         //Errors after lastRest

      private long totalMessages;  //Total messages
      private long totalErrors;        //Total errors
    }

We could just get away with 1 read and up to 1 write per integration per iteration.

# Record per reset
This is pretty similar to what's implemented already here, and goes like this:

- periodical query pods, grab data and create a single record (per integration & reset). Each time the reset value changes, we can create a new record.

Periodically we sum all metrics record, to create a sum record.

As I said this is pretty similar to what we have but I see two major differences:

- starting point for the rollups shouldn't be the integration, neither the pod, but the individual metrics records (which are going to be summed). It's what is called RawMetrics in this PR (if I get it right).

- The rollup should not be a different task, possibly handled by a different executor, definitely with a larger interval.

Reply:
@iocanel the first approach seems to be suitable if we are measureing relative numbers (since last read), the second one if we only get absolute (total) numbers from the MBeans. 

> The rollup should not be a different task, possibly handled by a different executor, definitely with a larger interval.

guess this should read 'should _be_ a different task' ? However, can't we make the sum up directly with the metrics query (as you can do for SQL or for Prometheus) ? This would completely eliminate the need for such a maintenance task.

Reply:
When pod get restarted the metrics gets reset. So we need to store periodically. Plus we need to add metrics from dead pod and live pod(s). It can be a different task yes, but I don't think the period of that should be too large since you will loose messages that weren't stored at the time the pod went down.

Reply:
At the moment for each tick we do the following:

1) read pod metrics
 i) write per integration
2) rollup pod metrics
 i) read all integrations and for each of them
ii) read all raw metrics
iii) write summary per integration.

What I suggest is to take the number of db operations down to a maximum 1  read/write for each pod (not integration).
if you just store the summary and you keep track of the totals and the counters since last reset, you don't need anything else.